### PR TITLE
Scheduler for M5CoreInk, secure upload, utilities, OTA firmware updates and secure boot

### PR DESCRIPTION
**Pull request by n-vr**
## Inside this pull-request
This pull-request adds several features:
- New scheduler for the M5CoreInk that puts the device in deep sleep.
- Secure upload that uploads everything inside a queue (still in volatile memory) to the backend API.
- Utility code.
- Over-the-air firmware updates (using GitHub, not the backend API).
- Secure boot (not enabled by default, but possible to use).

This pull-request is a big one. Don't worry. I prefer to keep them a lot smaller in the future. The following branches were merged into this one:
- main_m5coreink
- ota-firmware-update
- utilities
- secure boot

## Trello cards
https://trello.com/c/DW0lqhbK - Enable OTA firmware updates
https://trello.com/c/f4IwUp2j - refactor secure_upload code into separate lib folder
https://trello.com/c/x3m9N0t4 - improved support for root certificate (not in source code / use mbedtls)
https://trello.com/c/L09ZXwzY - Secure Upload component that does measurement timestamping, queueing (persistent if needed), (initiates) JSON formatting, upload timestamping and secure sending
https://trello.com/c/xsUQJlBh - Timer component that uses RTC+power down (or, of not available, deep sleep) to wake up the ESP32 and sensors when needed
https://trello.com/c/rn7yvcJO - Introduce toJSON() delegate that Secure Upload can call; use JSON library in <sensor>.toJSON() implementations
https://trello.com/c/cySrOKlY - fix logging of new_fw and booted_fw (not logged now)

### CODE REVIEW

File: lib/secure_upload/property_format.c [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    "%d", // PROPERTY_HEARTBEAT
+    "%u", // CO2_CONCENTRATION
+    "%.1f", // ROOM_TEMP_CO2
+    "%.1f" // HUMIDITY
+};
+
+// list that tells for each property its name in string format
+const char *name_of_property[] = {
+    "heartbeat",    // PROPERTY_HEARTBEAT
+    "CO2concentration", // CO2_CONCENTRATION
```
Comment (henriterhofte) [inline lib/secure_upload/property_format.c:15]: The bottom three properties measured by the SCD41 should ultimately not be in the generic firmware but in firmware repo of https://github.com/energietransitie/twutility-twomes-co2-bt-occupancy-monitor-firmware

File: lib/secure_upload/property_format.c [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#include "property_format.h"
+#include "upload.h"
+
+// list that tells for each property its format string
+const char *format_property[] = {
+    "%d", // PROPERTY_HEARTBEAT
+    "%u", // CO2_CONCENTRATION
```
Comment (henriterhofte) [inline lib/secure_upload/property_format.c:7]: The bottom three properties measured by the SCD41 should ultimately not be in the generic firmware but in firmware repo of https://github.com/energietransitie/twutility-twomes-co2-bt-occupancy-monitor-firmware

File: lib/secure_upload/property_format.c [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+void format_float(void *object, char *formated_value) {
+    measurement_t *obj = (measurement_t *) object;
+    sprintf(formated_value,format_property[obj->property],obj->value._float);
+}
+
+// list that tells for each property which function to use for formatting value to string
+void (*const format_function_of_property[])(void *, char *) = {
+    format_int,    // PROPERTY_HEARTBEAT
+    format_uint16, // CO2_CONCENTRATION
```
Comment (henriterhofte) [inline lib/secure_upload/property_format.c:39]: The bottom three properties measured by the SCD41 should ultimately not be in the generic firmware but in firmware repo of https://github.com/energietransitie/twutility-twomes-co2-bt-occupancy-monitor-firmware

**Review by henriterhofte — APPROVED** [review]
See my review comments. perhaps removing the stuff (including some library files) that are only needed for CO2-meter-SCD4x and that should only appear in https://github.com/energietransitie/twutility-twomes-co2-bt-occupancy-monitor-firmware is wise to do before we merge?

File: lib/secure_upload/property_format.c (hunk 2 above)
Comment (n-vr) [reply lib/secure_upload/property_format.c]: Fixed in https://github.com/energietransitie/twomes-generic-esp-firmware/commit/cfa8e2cbd446a736b5f27a06d15798c9525fe762

File: lib/secure_upload/property_format.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#ifndef PROPERTY_FORMAT_H
+#define PROPERTY_FORMAT_H
+
+// enum with all property types
+typedef enum {
+    PROPERTY_HEARTBEAT = 0,
+    CO2_CONCENTRATION,
```
[the remark this answers is not in the document]
Comment (n-vr) [reply lib/secure_upload/property_format.h]: Fixed in https://github.com/energietransitie/twomes-generic-esp-firmware/commit/cfa8e2cbd446a736b5f27a06d15798c9525fe762

File: lib/twomes_scd41/twomes_scd41.c [hunk 5]
```diff
@@ -0,0 +1,157 @@
+#include "twomes_scd41.h"
```
[the remark this answers is not in the document]
Comment (n-vr) [reply lib/twomes_scd41/twomes_scd41.c]: Fixed in https://github.com/energietransitie/twomes-generic-esp-firmware/commit/332969d6856b05883835603c935fb27d56f73aac

File: lib/twomes_scd41/twomes_scd41.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
+/**
```
[the remark this answers is not in the document]
Comment (n-vr) [reply lib/twomes_scd41/twomes_scd41.h]: Fixed in https://github.com/energietransitie/twomes-generic-esp-firmware/commit/332969d6856b05883835603c935fb27d56f73aac

File: platformio.ini [hunk 7]
```diff
@@ -8,15 +8,27 @@
...
+;Device and Platform settings:
+platform = espressif32@3.2.0
+board = m5stack-coreink
+framework = espidf
 
+[env]
 ;memory settings
 board_upload.flash_size=4MB
-board_upload.maximum_size=4194304
-board_build.partitions = no_ota.csv
```
[the remark this answers is not in the document]
Comment (n-vr) [reply platformio.ini:19]: Removed in https://github.com/energietransitie/twomes-generic-esp-firmware/commit/f91d285af2e0a036b3dc1f7a5ad8d774bea2849e

**Pull request merged**